File: 4.2.minimal-tree/solution.py
class Node:
    def __init__(self, data, left = None, right = None):
        self.data = data
        self.left = left
        self.right = right

    def __str__(self, level=0):
        ret = ""
        if self.right:
            ret += self.right.__str__(level + 1)
        ret += "    " * level + f"{self.data}\n"
        if self.left:
            ret += self.left.__str__(level + 1)
        return ret

class Solution:
    def minimal_tree(self, lst):
        return self.build_tree(lst)


    # Return the node perspective
    def build_tree(self, lst, left = 0, right = None):
        if right is None:
            right = len(lst) - 1
        if left > right:
            return None
        mid = (right + left) // 2
        n =  Node(lst[mid])
        n.left = self.build_tree(lst, left, mid - 1)
        n.right = self.build_tree(lst, mid + 1, right)
        return n



s = Solution()
rs = s.build_tree([1, 2, 3, 4, 5, 6, 7, 8, 9])
print(rs)

File: 4.2.minimal-tree/Solution.java
import java.util.*; 
class Solution { 
    public static void main (String [] args ){ 
        System.out.println("AC"); 
    }
    /**
     * Receive the tree and recurvie create the tree node 
     */
    public static class TreeNode<T extends Comparable<T>>{ 
        public T val; 
        public TreeNode<T> left; 
        public TreeNode<T> right; 
        public TreeNode(T val){ 
            this.val = val; 
        }
        public TreeNode(T val, TreeNode<T> left, TreeNode<T> right){ 
            this.val = val; 
            this.left = left; 
            this.right = right; 
        }



        public static <T extends Comparable<T>> TreeNode<T> buildTree(List<T> elements){ 
            if (elements == null){ 
                return null; 
            }
            return buildTree(elements, 0, elements.size() - 1);
        }

        public static <T extends Comparable<T>> TreeNode<T> buildTree(List<T> elements, int start, int end){ 
            if (elements == null){ 
                return null; 
            }
            if (start > end){ 
                return null; 
            }
            // Create the midddle node 

            int mid = (start + end) / 2; 
            TreeNode<T> n  = new TreeNode<>(elements.get(mid)); 
            var l = buildTree(elements, start, mid - 1); 
            var r = buildTree(elements, mid + 1, end); 
            n.left = l; 
            n.right = r; 
            return n; 
        }
    }
}
